integration/map.go

```diff
@@ -0,0 +1,416 @@
+	"github.com/google/trillian"
+
+	tcrypto "github.com/google/trillian/crypto"
+	stestonly "github.com/google/trillian/storage/testonly"
+
+	_ "github.com/google/trillian/merkle/coniks"
+	_ "github.com/google/trillian/merkle/maphasher"
+)
+
+var h2b = testonly.MustHexDecode
+
+// createBatchLeaves produces n unique map leaves.
+func createBatchLeaves(batch, n int) []*trillian.MapLeaf {
+	leaves := make([]*trillian.MapLeaf, 0, n)
+	for i := 0; i < n; i++ {
+		leaves = append(leaves, &trillian.MapLeaf{
+			Index:     testonly.TransparentHash(fmt.Sprintf("batch-%d-key-%d", batch, i)),
+			LeafValue: []byte(fmt.Sprintf("batch-%d-value-%d", batch, i)),
+		})
+	}
+	return leaves
+}
+
+func isEmptyMap(ctx context.Context, mCli trillian.TrillianMapClient, tree *trillian.Tree) error {
+	r, err := mCli.GetSignedMapRoot(ctx, &trillian.GetSignedMapRootRequest{
+		MapId: tree.TreeId,
+	})
+	if err != nil {
+		return fmt.Errorf("failed to get empty map head: %v", err)
+	}
+
+	if got, want := r.GetMapRoot().GetMapRevision(), int64(0); got != want {
+		return fmt.Errorf("got SMR with revision %d, want %d", got, want)
+	}
+	return nil
+}
+
+func verifyGetMapLeavesResponse(getResp *trillian.GetMapLeavesResponse, indexes [][]byte,
+	wantRevision int64, pubKey crypto.PublicKey, hasher hashers.MapHasher, treeID int64) error {
+	if got, want := len(getResp.MapLeafInclusion), len(indexes); got != want {
+		return fmt.Errorf("got %d values, want %d", got, want)
+	}
+	if got, want := getResp.GetMapRoot().GetMapRevision(), wantRevision; got != want {
+		return fmt.Errorf("got SMR with revision %d, want %d", got, want)
+	}
+
+	// SignedMapRoot contains its own signature. To verify, we need to create a local
+	// copy of the object and return the object to the state it was in when signed
+	// by removing the signature from the object.
+	smr := *getResp.GetMapRoot()
+	smr.Signature = nil // Remove the signature from the object to be verified.
+	if err := tcrypto.VerifyObject(pubKey, smr, getResp.GetMapRoot().GetSignature()); err != nil {
+		return fmt.Errorf("VerifyObject(SMR): %v", err)
+	}
+	rootHash := getResp.GetMapRoot().GetRootHash()
+	for _, incl := range getResp.MapLeafInclusion {
+		leaf := incl.GetLeaf().GetLeafValue()
+		index := incl.GetLeaf().GetIndex()
+		leafHash := incl.GetLeaf().GetLeafHash()
+		proof := incl.GetInclusion()
+
+		if got, want := leafHash, hasher.HashLeaf(treeID, index, leaf); !bytes.Equal(got, want) {
+			return fmt.Errorf("HashLeaf(%s): %x, want %x", leaf, got, want)
+		}
+		if err := merkle.VerifyMapInclusionProof(treeID, index,
+			leaf, rootHash, proof, hasher); err != nil {
+			return fmt.Errorf("VerifyMapInclusionProof(%x): %v", index, err)
+		}
+	}
+	return nil
+}
+
+// newTreeWithHasher is a test setup helper for creating new trees with a given hasher.
+func newTreeWithHasher(ctx context.Context, aCli trillian.TrillianAdminClient, hashStrategy trillian.HashStrategy) (*trillian.Tree, hashers.MapHasher, error) {
```

Comment:
I tend to call these clients something like `tadmin`.  Up to you if you think that's cleaner

Reply:
Done, in the interest of consistency.  Also 'admin' means more to people than 'a'.
Did the same for mCli -> tmap as well.